### PR TITLE
[RFC] Sub probing syzygy TB

### PR DESCRIPTION
**Pull request by mstembera**
bench: 6079565

This is a draft implementation to get useful information from syzygy TB for positions that have one too many pieces.  The idea is that if we remove one of our pieces and the resulting position is a TB win it's very likely the original position is also a win.  Conversely if we remove one of our opponents pieces and the resulting position is a TB loss very likely the original position is also a loss.   This isn't 100% true of course but hopefully often enough to gain elo.  The current version is slightly weaker than master but I don't know the best way to implement syzygy probing.   For example is there a way to probe using
```
template<TBType Type, typename Ret = typename TBTable<Type>::Ret>
Ret probe_table(const Position& pos, ProbeState* result, WDLScore wdl = WDLDraw)
```
directly w/o calling 
```
template<bool CheckZeroingMoves>
WDLScore search(Position& pos, ProbeState* result)
```
If anyone can look at the changes and offer comments or suggestions I would appreciate it.
@syzygy1 Yours would be especially helpful.

### CODE REVIEW

**Comment by vondele** [comment]
I wonder if you could use that as a heuristic more generally in search (i.e. not relate to TB probing), a bit like null-move pruning. If you remove a piece (e.g. the one that is moved in the TT move), and a reduced depth search returns a win, the position is almost certainly a win.

**Comment by dubslow** [comment]
viz ran some tests akin to vondele's statement a few months ago, tho without serious effort in writing the code and without serious results

**Comment by nickolasreynolds** [comment]
If the speed of Syzygy probing is holding this idea back, it may work nearly as well to remove only a single random piece (or the one with the lowest PSQT value, or some other quick-to-check heuristic), rather than exhaustively checking all possibilities. If the NNUE is still being used in close-to-tablebase positions, you could also ask it to recommend a piece for removal.

**Comment by mstembera** [comment]
@vondele That's an interesting idea.  The main difference is that in normal search you can always evaluate the exact position you care about which is presumably more efficient than estimating things based on similar positions.  In TB probing if a position has more pieces than you TB's contain you can't probe it at all.

@nickolasreynolds I will try your idea of just picking one piece to remove.  I don't however know how to ask NNUE to recommend a piece to remove?

**Comment by nickolasreynolds** [comment]
@mstembera By "ask the NNUE" I meant "add a new head to the architecture for recommending a piece to remove, retrain the best network for a bit, and see what happens." Using the NNUE might just be overkill, however, if you only intend to apply the removal heuristic to tablebase-adjacent positions.

Roping in the NNUE was mostly directed at how to make the removal heuristic apply more generally, as Vondele proposes. Perhaps it would allow the network to better learn and encode knowledge akin to, "Even if some of my unimportant material vanishes, I still seem to be doing well, so I may have a decisive advantage that I should press," and maybe that's a good perspective to have.

That sort of understanding seems to be an important component of human strategic thinking in many domains, where experts are adept at chunking together and walling off huge swathes of the decision tree, because they understand there are contexts where some things can be entirely ignored.

Admittedly, this is to some extent what move ordering and other search heuristics are already emulating, but there's more than one way to skin a cat, and this could be an interesting step towards NNUE-ifying the entire search algorithm.

**Comment by nickolasreynolds** [comment]
And if you're wondering what sort of training signal might make sense for a removal heuristic head, your loss function could be something like:

- If removing the recommended piece worsens your position, loss = square of the difference between the current evaluation and the piece removed evaluation.
- If removing the recommended piece improves your position, loss = whatever the maximum penalty is.

That way the network would learn to recognize when positions are robust to removal of material, and in some sense why, but would ignore "wishful" scenarios where removing material allows you to deliver a devastating check, or untrap a piece, or whatever.

**Comment by mstembera** [comment]
@nickolasreynolds So I pushed a version that only removes the least valuable piece for each side according to psqt.  It is still about 5 elo worse than master as before.  I also tried selecting the piece for removal using NNUE in a similar way to what we do starting on line 303 in evaluate_nnue.cpp but it performed even worse.  Changing the NNUE architecture is beyond my current ability due to lack of Python and Pytorch knowledge.  I agree the idea of having NNUE output a suggested best move for a position instead of just a score could be generally useful for search though.  I think some version of Alpha0 did this also.  I wonder what @Sopel97 thinks.  He may have even tried it already.

@vondele Upon further thought I like your idea even more than initially.  Since removing a piece could produce upper/lower bounds with shallow searches it could be more efficient for pruning than the actual position.  I hope one of our search experts can figure out how to make use of it.

On a related note, I tested master w/ 6-men vs master no TB and after 114K STC games the gain is only 3.0elo +-1.5elo.  I believe it was around 15 elo in the past.  I wonder if this is because our latest net is so much better at evaluating the TB positions?

**Comment by mstembera** [comment]
I'm closing this so it doesn't pollute the PR page but additional comments continue to be welcome.
